SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp

```diff
@@ -589,24 +603,30 @@ void UGlobalStateManager::QueryGSM(bool bRetryUntilAcceptingPlayers)
 		}
 		else
 		{
-			bool bNewAcceptingPlayers = GetAcceptingPlayersFromQueryResponse(Op);
+			bool bNewAcceptingPlayers = false;
+			int32 QuerySessionId = 0;
+			bool bQueryResponseSuccess = GetAcceptingPlayersAndSessionIdFromQueryResponse(Op, bNewAcceptingPlayers, QuerySessionId);
 
-			if (!bNewAcceptingPlayers && bRetryUntilAcceptingPlayers)
+			if (!bQueryResponseSuccess)
 			{
-				UE_LOG(LogGlobalStateManager, Log, TEXT("Not yet accepting new players. Will retry query for GSM."));
-				RetryQueryGSM(bRetryUntilAcceptingPlayers);
+				UE_LOG(LogGlobalStateManager, Error, TEXT("Failed to extract AcceptingPlayers and SessionId from GSM query response. Will retry query for GSM."));
+			}
+			else if (bNewAcceptingPlayers != bExpectedAcceptingPlayers ||
+					 QuerySessionId != ExpectedSessionId)
+			{
+				UE_LOG(LogGlobalStateManager, Log, TEXT("GlobalStateManager did not match expected query state. Will retry query for GSM."));
 			}
 			else
 			{
 				ApplyDeploymentMapDataFromQueryResponse(Op);
+				Callback.ExecuteIfBound();
```

Comment:
A cleaner way of doing this now we have the callback would probably be to not automatically retry the query based on `bExpectedAcceptingPlayers` and `bRetryUntilRecievedExpectedValues` but to instead just simply run the callback on `bQueryResponseSuccess` and then have the callback (in the net driver) make another query in the future using a timer. Just realised this. 

Won't block on this but it would clean this code up / make it easier to understand.